doc: design for `atlas pull` for the edx-platform and its plugins FC-0012

OmarIthawi opened the pull request:
### Atlas integration proposal

This is a design document which include a couple of decisions related to the `XBlockI18nService` and the way translations are going to be refactored using the [`openedx-atlas` new pull command](https://github.com/openedx/openedx-atlas).

Please review the decision document in this pull request by checking the file tab or using the updated preview from the branch:

 - Preview the [0019-oep-58-atlas-translations-design.rst](https://github.com/Zeit-Labs/edx-platform/blob/atlas-proposal/docs/decisions/0019-oep-58-atlas-translations-design.rst) design document
 - Checkout the draft code for this proposal: https://github.com/openedx/edx-platform/pull/31997

### TODO

 - [x] Prepare the proposal
 - [x] Rough testing to validate the code changes locally in devstack: https://github.com/openedx/edx-platform/pull/31997
 - [x] Remove the code
 - [x] Get the proposal reviewed
 - [x] Merge this pull request
 - [x] Open a new code-only pull request

## Deadline

The deadline for this design is September 22nd due in order to meet the overall deadline for the [FC-0012 project](https://openedx.atlassian.net/l/cp/XGS0iCcQ).


References
----------

This contribution is part of the [FC-0012 project](https://openedx.atlassian.net/l/cp/XGS0iCcQ) which is sparked by the [Translation Infrastructure update OEP-58](https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0058-arch-translations-management.html#specification).

Up-to-date project overview and details are available in the [Approach Memo and Technical Discovery: Translations Infrastructure Implementation](https://docs.google.com/document/d/11dFBCnbdHiCEdZp3pZeHdeH8m7Glla-XbIin7cnIOzU/edit#) document.

Join the conversation on [Open edX Slack #translations-project-fc-0012](https://openedx.slack.com/archives/C04R6TUJB7T).

Check the links above for full information about the overall project.

Internalization is being re-architected in Open edX Python, XBlock, Micro-frontend, and other projects. There are a number of immediately visible changes:
 - Remove source and language translations from the repositories, hence no `.json`, `.po` or `.mo` files will be committed into the repos.
 - Add standardized `make extract_translations` in all repositories
 - Push user-facing messages strings into [openedx/openedx-translations](https://github.com/openedx/openedx-translations/).
 - Integrate root repositories with [openedx/openedx-atlas](https://github.com/openedx/openedx-atlas/) to pull translations on build/deploy time

Breaking Changes
----------------

One of the primary goals of the project is to **avoid breaking changes**. If you noticed any suspicious code, please raise your concern. But before that, please know the strategy we're following to avoid breaking changes:

**For XBlocks, plugins and the edX Platform:**
 - The standard XBlock translations path must be `conf/locale`. Therefore, we are creating a link from `conf/locale` to `translations` so Atlas can find the path without disrupting the current way of having translations locally within the XBlocks.
 - [openedx-translations](https://github.com/openedx/openedx-translations) will have a related PR that adds the XBlock to the pipeline. This will not affect the current way of managing/updating translations
 - At the end of the project, a clear change log will be added and all translations will be handled by Atlas. Thus, the local translation will be removed from the repo within the version bump
 - A notification for the community will be published to ensure that everyone knows why translations directories are removed from all repos.
 - A feature flag will be created to use both `atlas` and non-atlas method of pulling translations in the edX Platform.

openedx-webhooks commented:
Thanks for the pull request, @OmarIthawi!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+--------------------------------------
+
+The `Translation Management update OEP-58`_ proposal has been merged with
+the following changes to the way translations are managed in Open edX:
```
```suggestion
The `Translation Management update OEP-58`_ proposal has been merged with
the following changes to the way translations are managed in the Open edX platform:
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+
+- Move Translation Files to the `openedx-translations repo`_
+- Add `Transifex GitHub App <https://github.com/apps/transifex-integration>`_
+  to openedx Organization
```
```suggestion
- Add the `Transifex GitHub App <https://github.com/apps/transifex-integration>`_
  to openedx Organization
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+- Add `Transifex GitHub App <https://github.com/apps/transifex-integration>`_
+  to openedx Organization
+- Connect the `openedx-translations repo`_ to the
+  `openedx-translations project`_
```
It's not clear from this what the `openedx-translations` project is/that it is on Transifex.

OmarIthawi replied:
done

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+  to openedx Organization
+- Connect the `openedx-translations repo`_ to the
+  `openedx-translations project`_
+- Copy Transifex's Translation Memory and Combine Translators
```
Couple things here:
* It's probably worth linking to Transifex docs for people unfamiliar with these terms
* It's not clear where things are being copied from and where they are being copied to

OmarIthawi replied:
done

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+  `openedx-translations project`_
+- Copy Transifex's Translation Memory and Combine Translators
+- Get Translations Back for Deployment/Development and introduce the new
+  `openedx-atlas`_ translation tool.
```
```suggestion
- Utilize `openedx-atlas`_ to pull translations for development/deployment.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+If you're new to the `Translation Management update OEP-58`_ proposal, please
+review it in addition to the
+`Approach Memo and Technical Discovery - Translations Infrastructure Implementation`_
+document before continuing.
```
My assumption is most people won't read through the entirety of OEP-58 or the approach memo. Instead of instructing people to review both in their entirety, it probably makes sense to link to sections within it when applicable.

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+`Approach Memo and Technical Discovery - Translations Infrastructure Implementation`_
+document before continuing.
+
+Current Architecture/Implementation for XBlocks and Plugins
```
I think using "pre OEP-58" instead of "Current" will make this a little more future proof (just in case things aren't updated in the most timely manner once everything fully switches over to the OEP-58 workflow)

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+process needs to happen. Somtimes the process takes a full month to
+complete such as in the example below:
+
+- `Added French (Canada) and Japanese - xblock-drag-and-drop-v2 #220`_
```
It might be nice to break this down a little, something like:

* Translations live in the repo
* Translations are packaged with the rest of the code when published to pypi

Pros:
* Translations are always there

Cons:
* Adding translations means adding a commit directly to the repo
  * This can mean a complex integration with Transifex
  * This can mean a lengthy manual PR review process

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+XBlockI18nService will be able to find the Python translations and use load
+them via the ``__init__`` method of the `XBlockI18nService`_ which finds
+the ``text.mo`` files and make it available to the ``edx-platform``
+during the execution of the XBlock.
```
```suggestion
The `XBlockI18nService`_ loads translations for all installed XBlocks/Plugins 
via its ``__init__`` method. This makes all the ``text.mo`` files available to ``edx-platform``
during the execution of the XBlock.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+
+As of September 2023, there is no centralized method to bundle JavaScript
+translations in XBlocks. Non-XBlock plugins lacks JavaScript translations
+support altogether.
```
```suggestion
As of September 2023, there is no centralized method to bundle JavaScript
translations in XBlocks. Non-XBlock plugins lack JavaScript translation
support altogether.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+
+The de-factor stadnard method for bundling JavaScript translations in XBlocks
+is to use ``web_fragment`` and load the translations as part of the XBlock
+frontend static files on every XBlock load.
```
```suggestion
The de-facto standard method for bundling JavaScript translations in XBlocks
is to use ``web_fragment`` and load the translations as part of the XBlock
frontend static files on every XBlock load.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+
+This design has it's trade-offs but it's been widely adopted and out of the
+scope of this proposal. Therefore, we'll improve on the existing design to
+load the ``atlas`` JavaScript translations instead of the bundled ones.
```
This could probably be simplified down to something like:

"OEP-58 does not change this structure, it just makes the necessary changes to pull translations from the openedx-translations repo via atlas instead of having them live in the xblock repo itself"

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+-------------------------------------------------------------
+
+Update the ``make pull_translations`` command to use the ``atlas pull``
+if the ``OPENEDX_ATLAS_PULL`` environment variable is set.
```
```suggestion
Update the ``make pull_translations`` command to use ``atlas pull``
if the ``OPENEDX_ATLAS_PULL`` environment variable is set.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+This mostly is a non-controversial change that has been done in other repos
+already.
+
+The next section is a little more intricate and requires more discussion.
```
I'm not sure if we need all of this here. It seems to be trying to cover a lot of different bases:
* example code
* justification of decision
* talking about the next section

I feel like simplifying this down to just be "We're going to use `atlas` in `make pull_translations` like we do in `course-discovery` and `frontend-app-learning`" might suffice.

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+the need for managing the translations in each XBlock.
+
+The XBlock translations is already stored in the `openedx-translations repo`_
+and is accessible by the `openedx-atlas`_ command-line.
```
```suggestion
Instead of storing translation files for each XBlock and Plugin in their respective repositories,
we will use `openedx-atlas`_ to pull them from the `openedx-translations repo`_.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+
+The non-XBlock plugins such as `edx-val`_ are out of the scope of this
+proposal and will be handled separately as stated in the :ref:`non-goals`
+section.
```
How will this work when not all repositories are ready to make the switch? It's not clear how we'll handle having some repos pull using atlas and some not.

OmarIthawi replied:
Agreed. We should allow fallbacking to the embedded translations even if the Waffle Switch is enabled.

brian-smith-tcril replied:
I'm not sure exactly what you mean by this. By "allow fallbacking" do you mean allowing per-repo/package configuration of if we should use atlas or not? I foresee situations where we have *some* translations available via `openedx-translations`/`openedx-atlas` but they haven't reached parity with what is included in the repo for one reason or another, in which case `atlas` would still be able to pull something and wouldn't trigger an automatic fallback.

OmarIthawi replied:
The case which I'm concerned about is non-openedx XBlocks which I think would need fallback. For example, https://github.com/eduNEXT/xblock-limesurvey has a `locale` directory but it's not included in the https://github.com/openedx/openedx-translations repository.

The XBlockI18NService should be able to detect that there's no translations being pulled by `atlas` into the `plugins-locale` directory, therefore should use the `django.mo` from the XBlock itself.

In the case of `@openedx` XBlocks this shouldn't be a concern because all the translations will be synced into the new project on a regular basis up until complete deprecation.

OmarIthawi replied:
@brian-smith-tcril I've removed the Waffle Switch section because I think backward compatilbity should be there even after OEP-58 has been fully implemented to accommodate for third-party pull requests.

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+    $ manage.py lms atlas_pull_plugins_translations --list
+    drag_and_drop_v2
+    done
+    eox_tenant
```
naming nit: i think this works better as `atlas_pull_plugin_translations`

"pull plugins translations" reads to me as "pull plugins *and* translations," where as "pull plugin translations" implies that we're pulling the *plugin* translations.

```suggestion
- ``manage.py lms atlas_pull_plugin_translations --list``: List all XBlocks and
  Plugins installed in the ``edx-platform`` virtual environment. This will
  list of Python *module names* (as opposed to git repository names) of the
  installed XBlocks and Plugins e.g.::

    $ manage.py lms atlas_pull_plugin_translations --list
    drag_and_drop_v2
    done
    eox_tenant
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+  This list doesn't include plugins that are bundled within the
+  ``edx-platform`` repository itself such as the Video XBlock, the ``capa``
+  module and others. The reason for this is that their translations are
+  already included in the ``edx-platform`` translations.
```
```suggestion
  This list doesn't include plugins that are bundled within the
  ``edx-platform`` repository itself. Translations for bundled plugins 
  are included in the ``edx-platform`` translation files.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+
+
+  It will pull from the `edx-platform-links`_ directory to create the
+  following file tree::
```
same naming nit: `atlas_pull_plugin_translations`

```suggestion
- ``manage.py lms atlas_pull_plugin_translations``: This command
  will pull translations for installed XBlocks and Plugins by module name::
  
    $ atlas pull \
        translations/edx-platform-links/drag_and_drop_v2/conf/locale:conf/plugins-locale/drag_and_drop_v2 \
        translations/edx-platform-links/done/conf/locale:conf/plugins-locale/done \
        translations/edx-platform-links/edx_proctoring/conf/locale:conf/plugins-locale/edx_proctoring


  Resulting in the following file tree::
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+  properly yet.
+- The directory name may change from ``edx-platform-links`` to
+  ``edx-platform-modules`` but this is out of the scope of this proposal and
+  have little to no impact on the rest of the proposal.
```
I think we can probably get away without having these notes in here. If we rename the edx-platform-links directory in the openedx-translations repo we just need to make sure to update the docs here to reflect that. The "may not work properly yet" line is something we'll want to remove shortly anyways as we get everything in place.

OmarIthawi replied:
```suggestion
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+  `translations/edx-platform-links/drag_and_drop_v2 directory`_.
+
+This update is already implemented in the `openedx-translations repo`_ as of
+`edx-platform-links PR #353`_ which includes the details of the changes.
```
```suggestion
The `openedx-translations repo`_ directory structure organizes translation files by
repository name. ``edx-platform`` has no record of plugin repository names. As of
`edx-platform-links PR #353`_, it is also possible to use ``atlas`` to pull translations by
module name.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+#. When ``ENABLE_ATLAS_TRANSLATIONS`` is enabled, the atlas translations will
+   be used which is located in the ``edx-platform`` in an the git-ignored
+   ``edx-platform/conf/plugins-locale/drag_and_drop_v2/ar/LC_MESSAGES/text.po``
+   path. This file pulled by the ``pull_plugins_translations`` command.
```
```suggestion
``get_python_locale_directory`` will support two modes:

#. When ``ENABLE_ATLAS_TRANSLATIONS`` is disabled, the XBlock translation files
   included in the ``pip`` packages will be used. For example, the
   ``lib/python3.8/site-packages/drag_and_drop_v2/translations/ar/LC_MESSAGES/text.po``
   path for the Drag and Drop XBlock.

#. When ``ENABLE_ATLAS_TRANSLATIONS`` is enabled, the translation files pulled by ``atlas``
   from the `openedx-translations repo`_ will be used. For example, the 
   ``edx-platform/conf/plugins-locale/drag_and_drop_v2/ar/LC_MESSAGES/text.po``
   path for the Drag and Drop XBlock.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+``django.js`` file.
+
+This function needs to be used by the XBlocks in an opt-in backward-compatible
+manner.
```
```suggestion
A ``get_javascript_locale_path`` method will be added to the ``XBlockI18nService`` to provide XBlocks the
appropriate path to ``django.js`` translation files. This will allow XBlocks to utilize legacy packaged translations
or ``atlas`` pulled translations depending on configuration settings.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+XBlocks as described in the :ref:`js-translations` section.
+
+The ``i18n_js_namespace`` property will eliminate the need to hardcode the
+namespace the `XBlock Makefile compile_translations rule`_.
```
When first reading this it wasn't clear how the namespace property was working, I think it might be clearer to have this in the compile js translations section.

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+           return None
+
+
+New ``compile_plugins_js_translations`` command
```
naming nit: maybe either `compile_plugin_js_translations` or `compile_js_plugin_translations`. If you have other ideas I'm all ears!

OmarIthawi replied:
```suggestion
New ``compile_js_plugin_translations`` command
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This command will loop over XBlock modules that has the ``i18n_js_namespace``
+property and compile the JavaScript translations.
```
```suggestion
This command will loop over XBlock modules that have the ``i18n_js_namespace``
property and compile the JavaScript translations.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+
+For example if the Drag and Drop XBlock has the ``i18n_js_namespace``
+the ``compile_plugins_js_translations`` command will execute the following
+commands::
```
```suggestion
For example, if the Drag and Drop XBlock has the ``i18n_js_namespace`` property,
the ``compile_plugins_js_translations`` command will execute the following
commands::
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This command will loop over XBlock modules that has the ``i18n_js_namespace``
+property and compile the JavaScript translations.
```
> and compile the JavaScript translations.

It's not clear from this documentation what compiling js translations implies. I'm not sure if information explaining that should live here, or if we should link out to something else that explains it, but it'd be nice to make that a little more clear somehow.

OmarIthawi replied:
I've linked to the [`compilejsi18n` command docs](https://django-statici18n.readthedocs.io/en/latest/commands.html#compilejsi18n).

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,423 @@
+virtual environment.
+
+This is a non-trivial task and appears to add more complexity than necessary
+for not much added benefit.
```
It'd be nice to elaborate a little bit here. I think the fact that the XBlocks/plugins won't be used outside the context of `edx-platform` makes it so having `edx-platform` manage the translations instead of the modules themselves helps explain the "not much added benefit" part.

OmarIthawi replied:
```suggestion
This is a non-trivial task and appears to add more complexity than necessary
due to the fact that XBlocks and plugins won't be used outside the context of ``edx-platform``.
```

brian-smith-tcril reviewed (commented):
Thank you so much for putting this together! I did a big review pass on the documentation. Overall it's looking really great! All of my suggestions are just suggestions. Hopefully some of them will make this clearer for people unfamiliar to OEP-58!

brian-smith-tcril commented:
@OmarIthawi it looks like this has a couple of pycodestyle errors https://github.com/openedx/edx-platform/actions/runs/6171614902/job/16750213428?pr=33166 and commits that don't pass the linter. I think it should be good to go once those get cleaned up and all the checks pass.

OmarIthawi committed:
docs: design for atlas pull for the edx-platform and its plugins

OmarIthawi commented:
> @OmarIthawi it looks like this has a couple of pycodestyle errors https://github.com/openedx/edx-platform/actions/runs/6171614902/job/16750213428?pr=33166 and commits that don't pass the linter. I think it should be good to go once those get cleaned up and all the checks pass.

I've fixed the commit lint PRs and moved all the code into another PR:

 - https://github.com/openedx/edx-platform/pull/31997

This should be ready for review.

brian-smith-tcril commented:
@OmarIthawi commitlint still failed, looks like you used `doc` instead of `docs`

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,363 @@
+
+The `XBlockI18nService implementation pull request`_ (2016) introduced the
+support for XBlock translations in ``edx-platform`` and has the full
+context of the implementation.
```
```suggestion
The `XBlockI18nService implementation pull request`_ (2016) introduced
support for XBlock translations in ``edx-platform`` and has the full
context of the implementation.
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,363 @@
+`LtiConsumerXBlock.student_view`_ methods.
+
+In order to separate the XBlock translations from the platform, it's isolated
+in a separate ``gettext`` name space. For example, the Drag and Drop XBlock
```
```suggestion
in a separate ``gettext`` namespace. For example, the Drag and Drop XBlock
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,363 @@
+
+- ``manage.py lms atlas_pull_plugin_translations --list``: List all XBlocks and
+  Plugins installed in the ``edx-platform`` virtual environment. This will
+  list of Python *module names* (as opposed to git repository names) of the
```
```suggestion
  list the Python *module names* (as opposed to git repository names) of the
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,363 @@
+XBlocks that may or may not be included in the `openedx-translations repo`_.
+Third-party XBlocks like the `Lime Survey XBlock`_ will benefit from this
+backward compatibility because it's not included in the
+`xblocks Transifex project`_.
```
```suggestion
Third-party XBlocks that are not included in the
`xblocks Transifex project`_, such as the `Lime Survey XBlock`_,
will benefit from this backwards compatibility. 
```

brian-smith-tcril commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,363 @@
+
+For example, the `Drag and Drop XBlock get_static_i18n_js_url`_ will need to
+be updated to support both the ``XBlockI18nService`` new
+``get_javascript_locale_path`` method and the namespace.
```
```suggestion
be updated to support the new ``XBlockI18nService``
``get_javascript_locale_path`` method and the namespace.
```

brian-smith-tcril reviewed (approved):
Thank you so much for getting this all written up! I left a couple small grammar nits/wording suggestions, but assuming those are addressed this LGTM!

OmarIthawi commented on docs/decisions/0019-oep-58-atlas-translations-design.rst:
```diff
@@ -0,0 +1,362 @@
+Status
+======
+
+Accepted
```
right?

brian-smith-tcril replied:
I believe so, yes.

OmarIthawi commented:
@jmbowman would you mind checking this design decision?

cc: @brian-smith-tcril

dianakhuang reviewed (approved):
This looks reasonable to me, and I'd like to see it go ahead so we can iron out the specifics. I think the only thing I'm concerned about is how operators (like edx.org) are supposed to handle plugins that are not installed by default, but maybe that will become clearer as it's implemented.

OmarIthawi commented:
> This looks reasonable to me, and I'd like to see it go ahead so we can iron out the specifics.

Thanks @dianakhuang. 

> I think the only thing I'm concerned about is how operators (like edx.org) are supposed to handle plugins that are not installed by default, but maybe that will become clearer as it's implemented.

The design is mostly concerned about the plugins which aren't installed by default. Embdedded plugins like the Discussion XBlock/XModule and the video XBlock has been and will be treated similar to native Open edX code like courseware and student modules.

Most of the design here is focused on both external plugins such as Drag and Drop XBlock, LTI Consumer XBlock, Scorm XBlock. The design is intended to be indfifferent to the method of installation such as: 

 - `base.in`
 - `github.in`
 - `private.txt`
 - manual `pip install ...`


I'm sure there's a blind spot somewhere, therefore backward compatibility will be maintained to avoid imposing a sudden change on operators.

rgraber commented:
Do you need someone from our team to merge this?

OmarIthawi commented:
> Do you need someone from our team to merge this?

That would be great @rgraber! I don't have merge access to this repository.

OmarIthawi commented:
> > Do you need someone from our team to merge this?
> 
> That would be great @rgraber! I don't have merge access to this repository.

@brian-smith-tcril if this is mergeable, would you mind merging it?

openedx-webhooks commented:
@OmarIthawi 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/33166&entry.1484655318=https://github.com/OmarIthawi&entry.752974735=2023-09-04+15:13&entry.1917517419=2023-10-25+17:36&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
